keeper-gradle-plugin/src/test/kotlin/com/slack/keeper/KeeperFunctionalTest.kt

```diff
@@ -0,0 +1,365 @@
+
+  companion object {
+    @JvmStatic
+    @Parameters(name = "useR8={0}")
+    fun data(): List<Array<*>> {
+      return listOf(
+          arrayOf(true),
+          arrayOf(false)
+      )
+    }
+  }
+
+  @Rule
+  @JvmField
+  val temporaryFolder = TemporaryFolder()
+
+  /**
+   * Basic smoke test. This covers the standard flow and touches on the following:
+   * - Variant configuration
+   * - Packaging of intermediate jars with compiled sources (including Kotlin sources)
+   * - Generation of inferred rules
+   * - Inclusion of generated rules in the final proguard configuration
+   */
+  @Test
+  fun standard() {
+    val (projectDir, proguardConfigOutput) = prepareProject(temporaryFolder)
+
+    val result = runGradle(projectDir, "assembleReleaseAndroidTest")
+    assertThat(result.task(":jarAndroidTestClassesForAndroidTestKeepRules")?.outcome).isEqualTo(
+        TaskOutcome.SUCCESS)
+    assertThat(result.task(":jarAppClassesForAndroidTestKeepRules")?.outcome).isEqualTo(
+        TaskOutcome.SUCCESS)
+    assertThat(result.task(":inferAndroidTestUsage")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
+
+    // Ensure the expected parameterized minifier ran
+    val expectedMinifier = if (useR8) "R8" else "Proguard"
+    assertThat(result.task(
+        ":transformClassesAndResourcesWith${expectedMinifier}ForReleaseAndroidTest")?.outcome).isEqualTo(
+        TaskOutcome.SUCCESS)
+
+    // Assert we correctly packaged app classes
+    val appJar = projectDir.generatedChild("app.jar")
+    val appClasses = ZipFile(appJar).readClasses()
+    assertThat(appClasses).containsAtLeastElementsIn(EXPECTED_APP_CLASSES)
+    assertThat(appClasses).containsNoneIn(EXPECTED_ANDROID_TEST_CLASSES)
+
+    // Assert we correctly packaged androidTest classes
+    val androidTestJar = projectDir.generatedChild("androidTest.jar")
+    val androidTestClasses = ZipFile(androidTestJar).readClasses()
+    assertThat(androidTestClasses).containsAtLeastElementsIn(EXPECTED_ANDROID_TEST_CLASSES)
+    assertThat(androidTestClasses).containsNoneIn(EXPECTED_APP_CLASSES)
+
+    // Assert we correctly generated rules
+    val generatedRules = projectDir.generatedChild("inferredUsage.pro")
+    assertThat(generatedRules.readText().trim()).isEqualTo(EXPECTED_GENERATED_RULES)
+
+    // Finally - verify our rules were included in the final minification execution.
+    // Have to compare slightly different strings because proguard's format is a little different
+    val expected = if (useR8) EXPECTED_GENERATED_RULES else EXPECTED_PROGUARD_CONFIG
+    assertThat(proguardConfigOutput.readText().trim()).contains(expected)
+  }
+
+  // TODO test cases
+  //  Transitive androidTest deps using transitive android deps (i.e. like IdlingResource)
+  //  multidex (zip64 use in jars)
+  //  Error cases
+  //    missing variants - but shouldn't be necessary after https://github.com/slackhq/keeper/issues/2
+  //    ordering - missing android jar message
+  //  Custom r8 versions
```

Comment:
This is left as a toe-hold for future tests. I think the IdlingResource transitives example can be done as part of the standard test. Multidex may be hard to test and I'm not sure it's worth it considering it's just to verify a zip flag we already set works.